### PR TITLE
Kill In The Event of Extreme Overload

### PR DESCRIPTION
**Pull request by blt**
This is a work-in-progress PR to integrate @jinnipark's work into mainline lager.  As of the issuing of this PR, the code does not fully work. In particular:

  * tests fail
  * restart of lager_event fails

This relates to Issue #324 (RIAK-1643) (RIAK-2068) (RIAK-2433).

### CODE REVIEW

**Comment by jinnipark** [comment]
Thank you for this update.
I lost track of my PR for a while.
I'll close mine.

**Comment by mrallen1** [comment]
This looks pretty good as a first pass. I am a little bit unsure how this will work for non-default sinks and it'd be nice to have a test case for that scenario. The timer:sleeps in the test code make me wary. I'd prefer to have something like

```erlang
wait_until(_Fun, 0) ->
    {error, too_many_retries};
wait_until(Fun, Retries) ->
     case Fun() of
         true -> ok;
         false -> timer:sleep(?SLEEP), wait_until(Fun, Retries-1)
     end.
```

**Comment by blt** [comment]
Agreed, this is a pretty quick integration and does need to take into
account the other sinks.

Next step is getting the existing tests going, repairing the approach to be
inline with your previous and tackling additional sinks.

That seems straightforward. Present concern I have is that the sink doesn't
appear to come back online correctly. Did not have much investigation time.
Should be fun to figure out.

**Comment by blt** [comment]
@jinnipark Thanks for doing the original work! :D

**Comment by blt** [comment]
Hmm, on my local host two test suites--in addition to the tests that fail already on 3.1.0--still fail:

  * lager_manager_killer_test
  * lager_handler_watcher

Run in isolation, these suites run with no issue. I'll continue to look into the failure to try and pick apart what's going on, but if someone else had a notion to take a look I'd appreciate it. 

Question: is it expected that 8 tests in lager_test_backend will fail? I've confirmed failure on master branch from a clean build, on my system.

**Comment by mrallen1** [comment]
If you're testing on erlang 18 then yes that's a known test fault we didn't fix because we're not regularly testing 18

There's a ticket open on this but I don't remember what it is.

**Comment by blt** [comment]
One test failure remaining, tested on Erlang 17.

**Comment by mrallen1** [comment]
For the record, #275 (RIAK-1886) is the known Erlang 18 test failures.

**Comment by mrallen1** [comment]
Thanks for pulling this forward Brian. I'll take a look at it tonight or tomorrow.

**Comment by blt** [comment]
Much appreciated! Going to get pulled off for a bit but I still have
ambitions to determine what's wrong with that final test.

**Comment by mrallen1** [comment]
Your branch makes the reinstall fail because the handle_info which processes the `reinstall_handler` message was removed.  You can see on line 128 where a timer sends the message to the handler to reinstall a handler which failed to start initially.  This diff below makes all tests pass.

```diff
mallen-rMBP:lager mallen$ git diff
diff --git a/src/lager_handler_watcher.erl b/src/lager_handler_watcher.erl
index 336ee06..3d13cd3 100644
--- a/src/lager_handler_watcher.erl
+++ b/src/lager_handler_watcher.erl
@@ -80,6 +80,9 @@ handle_info({gen_event_EXIT, Module, Reason}, #state{module=Module,
         ok
     end,
     {noreply, State};
+handle_info(reinstall_handler, #state{module=Module, config=Config, sink=Sink} = State) ->
+    install_handler(Sink, Module, Config),
+    {noreply, State};
 handle_info(reboot, State) ->
     lager_app:boot(),
     {noreply, State};
```

**Comment by blt** [comment]
Thanks @mrallen1!

**Comment by mrallen1** [comment]
Cool. So the only remaining thing I'd like to see is a test with the killer that uses an alternative sink, then I think we'll be ready to merge this.

**Comment by mrallen1** [comment]
btw - not necessarily nominating you to write that test @blt - just more of a TODO item for myself.

File: src/lager_app.erl [hunk 1]
```diff
@@ -200,7 +221,10 @@ get_env_default({ok, Value}, _Default) ->
 
 start(_StartType, _StartArgs) ->
     {ok, Pid} = lager_sup:start_link(),
+    SavedHandlers = boot(),
+    {ok, Pid, SavedHandlers}.
 
+boot() ->
```
Comment (mrallen1) [inline src/lager_app.erl]: Let's make `boot/1` that takes a Sink argument. `boot/0` should call `boot/1` with the argument `?DEFAULT_SINK`

File: src/lager_manager_killer.erl [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+-include("lager.hrl").
+
+-record(state, {
+          killer_hwm :: non_neg_integer(),
+          killer_reinstall_after :: non_neg_integer()
+         }).
+
+kill_me() ->
+    gen_event:call(lager_event, ?MODULE, kill_self).
```
Comment (mrallen1) [inline src/lager_manager_killer.erl]: This needs to be abstracted to use a generic Sink name.

**Comment by mrallen1** [comment]
Just thought I'd update things. I spent most of yesterday and today working on this feature - generalizing it to multiple sinks, overhauling the start up code and so forth. I also spent some time with @macintux tracking down a race condition which revealed itself during testing. I'm currently seeing test failures in a single test and the failure is confounding me - I can't *quite* figure out why its failing, so I'm going to take a look at it tomorrow hopefully with fresh eyes and see if I can't figure it out.

When I get the last test functional I'm going to open a new PR which incorporates most of this PR and supersedes it.

**Comment by blt** [comment]
Awesome, thank you for doing the work to push this over the line! This is
going to be really great.

**Comment by mrallen1** [comment]
Closing in favor of #346

**Pull request closed without merging**